may-2018-reenabled-opcodes.md

```diff
@@ -0,0 +1,428 @@
+5. `x OP_0 OP_CAT -> x`. Concatenating an empty array onto any operand results in the operand, including when 
+   `len(x) = MAX_SCRIPT_ELEMENT_SIZE`
+6. `OP_0 x OP_CAT -> x`. Concatenating any operand onto an empty array results in the operand, including when 
+   `len(x) = MAX_SCRIPT_ELEMENT_SIZE`
+7. `x y OP_CAT -> concat(x,y)`. Concatenating two operands generates the correct result
+
+### OP_SPLIT
+
+*`OP_SPLIT` replaces `OP_SUBSTR` and uses it's opcode.*
+
+    Opcode (decimal): 127
+    Opcode (hex): 0x7f
+
+
+Split the operand at the given position.  This operation is the exact inverse of OP_CAT
+
+    x n OP_SPLIT -> x1 x2
+
+    where n is interpreted as a numeric value
+
+Examples:
+* `0x001122 0 OP_SPLIT -> OP_0 0x001122`
+* `0x001122 1 OP_SPLIT -> 0x00 0x1122`
+* `0x001122 2 OP_SPLIT -> 0x0011 0x22`
+* `0x001122 3 OP_SPLIT -> 0x001122 OP_0`
+
+Notes:
+* this operator has been introduced as a replacement for the previous `OP_SUBSTR`, `OP_LEFT`and `OP_RIGHT`. All three operators can be
+simulated with varying combinations of `OP_SPLIT`, `OP_SWAP` and `OP_DROP`.  This is in keeping with the minimalist philosophy where a single
+primitive can be used to simulate multiple more complex operations.
+* `x` is split at position `n`, where `n` is the number of bytes from the beginning
+* `x1` will be the first `n` bytes of `x` and `x2` will be the remaining bytes 
+* if `n == 0`, then `x1` is the empty array and `x2 == x`
+* if `n == len(x)` then `x1 == x` and `x2` is the empty array.
+* if `n > len(x)`, then the operator must fail.
+* `x n OP_SPLIT OP_CAT -> x`, for all `x` and for all `0 <= n <= len(x)`
+    
+The operator must fail if:
+* `!isnum(n)`. Fail if `n` is not a numeric value.
+* `n < 0`. Fail if `n` is negative.
+* `n > len(x)`. Fail if `n` is too high.
+
+Impact of successful execution:
+* stack memory use is constant (slight reduction by `len(n)`)
+* number of elements on stack is constant
+
+Unit tests:
+* `OP_0 0 OP_SPLIT -> OP_0 OP_0`. Execution of OP_SPLIT on empty array results in two empty arrays.
+* `x 0 OP_SPLIT -> OP_0 x`
+* `x len(x) OP_SPLIT -> x OP_0`
+* `x (len(x) + 1) OP_SPLIT -> FAIL`
+
+## Bitwise logic
+
+The bitwise logic operators expect 'byte sequence' operands. The operands must be the same length. 
+* In the case of 'byte sequence' operands `OP_CAT` can be used to pad a shorter byte sequence to an appropriate length.
+* In the case of 'byte sequence' operands where the length of operands is not known until runtime a sequence of 0x00 bytes 
+  (for use with `OP_CAT`) can be produced using `OP_0 n OP_NUM2BIN`
+* In the case of numeric value operands `x n OP_NUM2BIN` can be used to pad a numeric value to length `n` whilst preserving 
+  the sign bit.
+
+### OP_AND
+
+    Opcode (decimal): 132
+    Opcode (hex): 0x84
+
+Boolean *and* between each bit in the operands.
+
+	x1 x2 OP_AND -> out
+
+Notes:
+* where `len(x1) == 0 == len(x2)` the output will be an empty array.
```

Comment:
Although I understood this, it took a second. It might be clearer to express as "where len(x1) == 0 and len(x2) == 0".

Reply:
done